File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda x : (x[0], x[1]))
        res = []
        i = 1
        start = intervals[0][0]
        end = intervals[0][1]
        while i < len(intervals):
            if intervals[i][0] > end:
                res.append([start, end])
                start = intervals[i][0]
                end = intervals[i][1]
            else:
                end = max(end, intervals[i][1])                
            i += 1

        res.append([start, end])

        return res

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        // sort by start time
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        int longestEnd = 0;        
        int[] curr = intervals[0];
        List<int[]> merged = new ArrayList<>();
        for (int[] interval : intervals) {
            if (interval[0] <= curr[1]) {
                curr[1] = Math.max(curr[1], interval[1]);
            } else {
                merged.add(curr);
                curr = interval;
            }
        }
        merged.add(curr);
        int[][] mergedArray = new int[merged.size()][2];
        int i = 0;
        for (int[] interval : merged) {
            mergedArray[i++] = interval;
        }
        return mergedArray;
    }
}